usr/src/uts/common/fs/zfs/dsl_scan.c

```diff
@@ -1891,3 +2463,1073 @@ dsl_scan_restarting(dsl_scan_t *scn, dmu_tx_t *tx)
+		/*
+		 * Only remove what's consumed. The ss_fill field will have
+		 * already been adjusted above, so when the range tree
+		 * internally recycles the value in dsl_scan_queue_insert_cb
+		 * and dsl_scan_queue_remove_cb, it will work out OK.
+		 */
+		qzio = list_tail(list);
+		/*
+		 * We stick a zero-length qzio into q_adding_zio, so that
+		 * dsl_scan_queue_remove_cb saves the ss_fill field of the
+		 * old scan seg (which has already been decremented by the
+		 * amount we lifted from it in the loop above) and will be
+		 * restored in the new scan seg representing the new extent
+		 * tail.
+		 */
+		queue->q_adding_zio = zero_qzio;
+		range_tree_remove(queue->q_exts_by_addr, ss->ss_rs.rs_start,
+		    (QZIO_GET_OFFSET(qzio) + QZIO_GET_ASIZE(qzio)) -
+		    ss->ss_rs.rs_start);
+		kmem_free(zero_qzio, sizeof (*zero_qzio));
+	} else {
+		/*
+		 * Whole extent consumed, remove it all, including any head
+		 * or tail overhang.
+		 */
+		range_tree_remove(queue->q_exts_by_addr, ss->ss_rs.rs_start,
+		    ss->ss_rs.rs_end - ss->ss_rs.rs_start);
+	}
+	atomic_add_64(&queue->q_num_issuing_zios, num_zios);
+
+	return (bytes);
+}
+
+/*
+ * This is called from the queue emptying thread and selects the next
+ * extent from which we are to issue qzio's. The behavior of this function
+ * depends on the state of the scan, the current memory consumption and
+ * whether or not we are performing a scan shutdown.
+ * 1) We select extents in an elevator algorithm (LBA-order) if:
+ *	a) the scan has finished traversing metadata (DSS_FINISHING)
+ *	b) the scan needs to perform a checkpoint
+ *	c) the scan is performing an early shutdown
+ * 2) We select the largest available extent if we are up against the
+ *	memory limit.
+ * 3) Otherwise we don't select any extents.
+ */
+static scan_seg_t *
+dsl_scan_fetch_ext(dsl_scan_queue_t *queue, boolean_t shutdown)
+{
+	dsl_scan_t *scn = queue->q_scn;
+
+	ASSERT(MUTEX_HELD(&queue->q_vd->vdev_scan_queue_lock));
+	ASSERT0(queue->q_issuing_rs.rs_start);
+	ASSERT0(queue->q_issuing_rs.rs_end);
+	ASSERT(scn->scn_is_sorted);
+
+	if (scn->scn_phys.scn_state == DSS_FINISHING ||
+	    scn->scn_checkpointing || shutdown) {
+		/*
+		 * When the scan has finished traversing all metadata and is
+		 * in the DSS_FINISHING state or a checkpoint has been
+		 * requested, no new extents will be added to the sorting
+		 * queue, so the way we are sorted now is as good as it'll
+		 * get. So instead, switch to issuing extents in linear order.
+		 */
+		return (range_tree_first(queue->q_exts_by_addr));
+	} else if (scn->scn_clearing) {
+		return (avl_first(&queue->q_exts_by_size));
+	} else {
+		return (NULL);
+	}
+}
+
+/*
+ * Empties a scan queue until we have issued at least info->qri_limit
+ * bytes, or the queue is empty. This is called via the scn_taskq so as
+ * to parallelize processing of all top-level vdevs as much as possible.
+ */
+static void
+dsl_scan_queues_run_one(queue_run_info_t *info)
```

Comment:
Ok, I'll rewrite the queue handling so that we create a taskq with "ncpus" worth of threads and then divide the vdevs we need to handle evenly between them.